app/models/work.rb

```diff
@@ -183,13 +183,34 @@ def new_recipients_allow_gifts
     self.new_gifts.each do |gift|
       next if gift.pseud.blank?
       next if gift.pseud&.user&.preference&.allow_gifts?
-      next if self.challenge_assignments.map(&:requesting_pseud).include?(gift.pseud)
-      next if self.challenge_claims.reject { |c| c.request_prompt.anonymous? }.map(&:requesting_pseud).include?(gift.pseud)
+      next if challenge_bypass(gift)
 
       self.errors.add(:base, ts("%{byline} does not accept gifts.", byline: gift.pseud.byline))
     end
   end
 
+  validate :new_recipients_have_not_blocked_gift_giver
+  def new_recipients_have_not_blocked_gift_giver
+    return if self.new_gifts.blank?
+
+    self.new_gifts.each do |gift|
+      next if challenge_bypass(gift)
+
+      blocked_users = gift.pseud&.user&.blocked_users || []
+      next if blocked_users.empty?
+
+      pseuds_after_saving.each do |pseud|
+        next unless blocked_users.include?(pseud.user)
+
+        if User.current_user == pseud.user
+          self.errors.add(:base, ts("%{byline} does not accept gifts from you.", byline: gift.pseud.byline))
+        else
+          self.errors.add(:base, ts("%{byline} does not accept gifts.", byline: gift.pseud.byline))
```

Comment:
Would be nice to make these translatable. E.g.
```suggestion
          errors.add(:base, :blocked_your_gifts, byline: gift.pseud.byline)
        else
          errors.add(:base, :blocked_gifts, byline: gift.pseud.byline)
```
with the translations at `activerecord.errors.models.work.blocked_your_gifts` and `activerecord.errors.models.work.blocked_gifts`:

```yml
blocked_gifts: "%{byline} does not accept gifts."
blocked_your_gifts: "%{byline} does not accept gifts from you."
```

Extra nice would be to also do this for the error in `new_recipients_allow_gifts`.